### PR TITLE
Don't allow {} to refer to implicit captures in format_args.

### LINKED ISSUE #93378: Named format arguments introduce implicit positional arguments
I tried this code:

```rust
fn main() {
    let a = "a";
    let b = "b";
    
    println!("{a} {b} {} {} {c} {}", c = "c");
}
```

I expected to see a compile error that says that there are missing positional arguments or that it is prohibited to mix positional and named parameters.

Instead, this compiles and prints the following:
```
a b c a c b
```

This code is [available on the playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=91fcd6600665ef45dae25f2a735768fb).

### Meta

`rustc --version --verbose`:
```
rustc 1.58.1 (db9d1b20b 2022-01-20)
binary: rustc
commit-hash: db9d1b20bba1968c1ec1fc49616d4742c1725b4b
commit-date: 2022-01-20
host: x86_64-unknown-linux-gnu
release: 1.58.1
LLVM version: 13.0.0
```

### PR DESCRIPTION
**Pull request by m-ou-se**
Fixes #93378

### CODE REVIEW

**Comment by m-ou-se** [comment]
This could probably also use a test for formatting placeholders like `{:.*}`, where the width/precision refers to an implicit argument.

**Comment by CAD97** [comment]
This should imho probably also add an explicit test for the `println!("{a} {}", a=a)` case, to document it at least.

File: compiler/rustc_builtin_macros/src/format.rs [hunk 1]
```diff
@@ -369,7 +377,7 @@ impl<'a, 'b> Context<'a, 'b> {
 
         let count = self.pieces.len()
             + self.arg_with_formatting.iter().filter(|fmt| fmt.precision_span.is_some()).count();
-        if self.names.is_empty() && !numbered_position_args && count != self.args.len() {
+        if self.names.is_empty() && !numbered_position_args && count != self.num_args() {
```
Comment (CAD97) [inline compiler/rustc_builtin_macros/src/format.rs:380]: Normally this (positive) branch would result in a message `1 positional argument in format string, but no arguments were given`, whereas now adding an implicit capture gets `invalid reference to positional argument 0 (no arguments were given)`. Ideally this wording would be kept, even with implicit args; I _think_ the proper check would then be `self.names.len() == self.num_captured_args && !numbered_position_args && count != self.num_args()`?

Comment (m-ou-se) [reply compiler/rustc_builtin_macros/src/format.rs:380]: Then the `let count =` above would also have to be adjusted to exclude all placeholders that refer to captures, which is a bit more complicated.

I agree the diagnostics here can be improved when there's captured arguments involved, but I rather leave that to a separate change that can fix all those cases properly.

**Comment by m-ou-se** [comment]
Marked as beta-accepted per discussion in the libs api team just now.

**Comment by estebank** [comment]
r=me, but we really want a crater run of this to make sure we're not regressing here and am slightly uncomfortable with a beta backport (but it's not my call)

File: src/test/ui/fmt/format-args-capture-issue-93378.stderr [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+   |
+   = note: positional arguments are zero-based
+
+error: invalid reference to positional argument 0 (no arguments were given)
+  --> $DIR/format-args-capture-issue-93378.rs:9:23
+   |
+LL |     println!("{a:.n$} {b:.*}");
+   |               ------- ^^^--^
+   |               |          |
+   |               |          this precision flag adds an extra required argument at position 0, which is why there are 3 arguments expected
```
Comment (estebank) [inline src/test/ui/fmt/format-args-capture-issue-93378.stderr:15]: This output needs some work to make it clearer :(

Comment (m-ou-se) [reply src/test/ui/fmt/format-args-capture-issue-93378.stderr:15]: Yes, there's quite a few cases right now where this macro produces suboptimal diagnostics.

We have plans to change how `fmt::Arguments` is implemented/represented, which would require a lot of changes to the `format_args` builtin macro. That would probably be the best time to just rewrite the macro and the way it produces diagnostics.

**Comment by m-ou-se** [comment]
> we really want a crater run of this to make sure we're not regressing here

We could run this PR through crater, but I suppose the beta crater run is good enough to catch any problems.

**Comment by m-ou-se** [comment]
Looks like the crater queue is almost empty, so we can quickly run this through crater to be sure:

@craterbot check

**Comment by m-ou-se** [comment]
I was told crater needs some maintenance after [the current experiment](https://crater.rust-lang.org/ex/pr-91743) finishes, so let's hold off until that's done.

@pietroalbini can you start a `check` run here once you're done with that? Thanks.

**Comment by apiraino** [comment]
T-compiler discussion about beta-backport approval [on Zulip](https://zulip-archive.rust-lang.org/stream/238009-t-compiler/meetings/topic/.5Bweekly.5D.202022-02-03.20.2354818.html#270565450)

**Comment by apiraino** [comment]
T-compiler discussion about beta-backport approval [on Zulip](https://zulip-archive.rust-lang.org/stream/238009-t-compiler/meetings/topic/.5Bweekly.5D.202022-02-03.20.2354818.html#270565450)

Also, stable backport nomination discussed and approved ([Zulip](https://zulip-archive.rust-lang.org/stream/238009-t-compiler/meetings/topic/.5Bweekly.5D.202022-02-03.20.2354818.html#270566882)), in particular that this backport should be _included_ in a point release, but does not _motivate_ one.

@rustbot label +stable-nominated +stable-approved

**Comment by Mark-Simulacrum** [comment]
@craterbot run mode=check-only start=master#1be5c8f90912c446ecbdc405cbc4a89f9acd20fd end=try#1b0139f5bdae7328e6d232947a8112333bb44eeb cap-lints=warn

**Comment by m-ou-se** [comment]
Just two failures in the crater run. Both of them are unrelated: a build script failing to download some file.

@bors r=estebank

**Pull request merged**